File: python/backtracking/word_search.py
from typing import List

# 单词搜索


class Solution:

    def exist(self, board: List[List[str]], word: str) -> bool:
        visited = [[False] * len(board[0]) for _ in range(len(board))]
        for i in range(len(board)):
            for j in range(len(board[i])):
                if self.helper(board, word, visited, i, j, 0):
                    return True
        return False

    def helper(self, board, word, visited, i, j, k) -> bool:
        if k == len(word):
            return True
        if i < 0 or j < 0 or i >= len(board) or j >= len(board[0]):
            return False
        if board[i][j] != word[k] or visited[i][j]:
            return False

        visited[i][j] = True
        for x, y in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
            if self.helper(board, word, visited, i + x, j + y, k + 1):
                return True
        visited[i][j] = False
        return False

File: go/backtracking/word_search.go
package main

//单词搜索

func exist(board [][]byte, word string) bool {
	visited := make([][]bool, len(board))
	for i := range visited {
		visited[i] = make([]bool, len(board[0]))
	}
	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[i]); j++ {
			if existHelper(board, i, j, 0, word, visited) {
				return true
			}
		}
	}
	return false
}

func existHelper(board [][]byte, i, j, k int, word string, visited [][]bool) bool {
	if k == len(word) {
		return true
	}
	if i < 0 || j < 0 || i >= len(board) || j >= len(board[0]) {
		return false
	}
	if board[i][j] != word[k] || visited[i][j] {
		return false
	}

	visited[i][j] = true
	if existHelper(board, i-1, j, k+1, word, visited) || existHelper(board, i, j-1, k+1, word, visited) || existHelper(board, i+1, j, k+1, word, visited) || existHelper(board, i, j+1, k+1, word, visited) {
		return true
	}
	visited[i][j] = false
	return false
}
